### ISSUE TITLE
Placeholders should come from Bower

### ISSUE THREAD

**Issue by PaulL1** [ngbp/ngbp#120]
I'm having a problem where I lose the vendor/placeholder directory.

The problem looks to be that the .gitignore ignores everything in the vendor/ directory, but the npm install and other commands don't recreate the vendor/placeholder directory.  

My use case is that I:
1. Clone the ng-boilerplate repository
2. Clobber the existing .git directory, and then run git init to create a new git
3. Push it to my git server
4. Check out onto another machine

At this point, that other machine appears to have no way to get or create the vendor/placeholder directory.

I'm pretty new to angular (and, to be honest, to javascript and nodejs as well).  So I don't have a great suggestion as to a fix - but options I can see are:
1. I shouldn't have clobbered the base git, and should have just forked it.  I'm not really clear that I want my new project to be a fork of this project, nor that you'd want someone to accidentally push changes from my project back into your upstream.  Perhaps my usage and understanding of git is flawed?
2. I manually push the vendor/placeholders directory into my local git (using "git add -f vendor/*".  This is my current solution, which works fine, the main issue is that's not intuitive to others.  If this is the right answer, then perhaps the documentation should highlight it as something to do.
3. Perhaps the npm install or bower install should be capable of recreating the placeholder directory - so the expectation is that when you checkout to a new machine you run npm install and bower install, and that should result in a populated vendor directory

I suspect option 3 is the right one, but I'm concerned that maybe option 1 is my problem.

**Comment by joshdmiller** [comment]
@PaulL1 Thank you for writing such a thorough issue! Too often, I see GitHub issues as "I can't ___", which gives us little to go on. I very much appreciate the thoughtfulness! :-)

First, a little background - the `vendor/` directory is in the `.gitignore` because that directory is _supposed_ to be just for Bower components; that said, someone made the unfortunate mistake of registering my placeholders library with Bower without using a built version! The process for the Bower team to correct this is manual (or I could use a non-canonical name, which is silly) and so I opted to just stick it in there and manually add the directory (with `git add vendor/placeholders` - you shouldn't add the whole `vendor/` folder) as a temporary solution. So that said, the best long-term solution is sticking to the Bower components and fixing this upstream. I have thus renamed this issue and will keep it open with the intention of fixing it.

As an aside, the clobbering method is probably not the best approach for starting new ngBoilerplate projects; or, at least it's not the canonical method when it comes to git. I understand not wanting to use a fork because it forces you to keep the commit history (and, for the record, when I start new projects I don't fork it either). I really should add to the documentation how to easily "fork" ngBoilerplate without the commit history, but it's actually fairly straightforward. First, create your own empty repository:

``` sh
$ git init .
$ touch README.md
$ git add README.md
$ git commit -m "chore(*): Initial commit"
```

Now add ngBoilerplate as a remote:

``` sh
$ git remote add ngboilerplate git://github.com/joshdmiller/ng-boilerplate.git
$ git fetch ngboilerplate # this doesn't touch your code; it just adds the git history, etc.
$ git merge --squash ngboilerplate/v0.3.0-release # this pulls in the code, but not the history
```

At this point, all the files are merged in, but the commit history is "squashed"; you now need to commit the changes, which combines _all_ the ngBoilerplate commits from the old repo you just merged into a single commit on your new repo:

``` sh
$ git commit -m "chore(*): Added ngBoilerplate"
$ git log --format=oneline
blahblahblah chore(*): Initial commit
blahblahblah chore(*): Added ngBoilerplate
```

And that's it. Your new repository has all the ngBoilerplate goodness, but none of the commit history. I think this is the more canonical method of starting new projects. Perhaps what I should do is script it into a little NPM-installable CLI. Something like `npm -g install ngboilerplate` and then `ngboilerplate new` to do it all for us...

Anyway, until this is fixed, you'll have to `git add vendor/placeholders` manually. But we'll get it fixed. :-)

**Comment by PaulL1** [comment]
@joshdmiller: thanks in return for the thorough, and very rapid, reply.  I understand the suggestion of add ngboilerplate as a remote, and that gives exactly what I want without causing this issue.  Agree that dropping that into the documentation somewhere would probably make sense, but also probably not urgent if there's a long-term resolution coming that avoids the issue entirely.

**Comment by umurkontaci** [comment]
@joshdmiller I find the git trick very useful, until a script is made for this, I think you can put this snippet somewhere in the readme or attach as a gist. I believe it would ease a lot people's job.

**Comment by mboudreau** [comment]
I agree with this.  I just spent an hour trying to figure out why it worked fine on my computer (I setup the project) while all my coworkers couldn't get it working.  This should be removed as it adds a manual dependency in a folder that should be ignored by git, and secondly, why the hell is it even in the project?!

**Comment by joshdmiller** [comment]
@mboudreau A kickstarter can't prevent misuse of the underlying tools. The problem is encountered because both git and bower were used incorrectly. The documentation could use a note, perhaps, to help explain what's going on (and there's an open issue to add this to the docs). A PR would be welcome too. :-) 

And the placeholders are included because this is a kickstarter. The first step of most new projects is design, and "lorem ipsum" text and placeholder images are very helpful to me when doing that. It can always be removed immediately, so I don't think it's a big deal that it's there. That said, if the community at large wants it gone, we can remove it.

**Comment by jamiebarrow** [comment]
Thanks for the explanation @joshdmiller, I never knew about the `git merge --squash` option, one question, what does the `chore(*):` part do? Is it something special, that runs some kind of command, or is it just some convention you have with your commit comments? Sorry if this is off topic.

**Comment by joshdmiller** [comment]
@jamiebarrow It's just a strategy for organizing commits. The outer part indicates the type of commit ("chore", "feat", "fix", etc.) and the inner part indicates the feature on which the commit is acting; the asterisk indicates no particular feature or impacting all features.

In and of itself, this does nothing. But it does two things: first, it makes commit logs easier to read; and second, it allows for some pretty nifty automation, like the generation of changelogs. The changelog in this project is automatically generated by parsing these commit messages (and then subsequently edited for clarity).

This strategy is used by many projects, including AngularUI, Angular Widgets (ui-bootstrap), and AngularJS itself. You can read more [here](https://docs.google.com/document/d/1QrDFcIiPjSLDn3EL15IJygNPiHORgU1_OOAqWjiDU5Y/edit#).

**Comment by dholbrook** [comment]
My workaround:

1) remove /vendor from the root .gitignore

2) create a new .gitignore in the vendor directory

```
# Ignore everything:
*

# Except for placeholders

!.gitignore
!placeholders/
!placeholders/*
```

**Comment by wcravens** [comment]
I just added `vendor/*` && `!vendor/placeholders` to the root level `.gitignore`.  I think it better represents the issue.

This also allows me to add future `!vendor/foo` entries.  If you leave `vendor/` then git will continue to ignore the subdirectories even with the negated entry.
